oak_functions/loader/src/server.rs

```diff
@@ -0,0 +1,118 @@
+//
+// Copyright 2021 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use anyhow::Context;
+use http::{request::Request, response::Response};
+use hyper::{
+    service::{make_service_fn, service_fn},
+    Body, Server, StatusCode,
+};
+use log::info;
+use std::{fs, net::SocketAddr, sync::Arc};
+
+const MAIN_FUNCTION_NAME: &str = "main";
+
+/// An HTTP server that serves the `main` function of a wasm module on each invocation.
+pub struct WasmServer {
+    /// Server address to listen for client requests on.
+    address: SocketAddr,
+    /// The wasm engine that runs the given wasm module per incoming request.
+    wasm_engine: WasmEngine,
+}
+
+#[derive(Clone)]
+struct WasmEngine {
+    // wasm module to be served on each invocation.
+    module: Arc<wasmi::Module>,
+}
+
+impl WasmEngine {
+    async fn handle_request(&self, req: Request<Body>) -> anyhow::Result<Response<Body>> {
```

Comment:
Could you add a simple benchmark for this? I would be curious to see how long it actually takes to start the new wasm engine each time. Can be a separate PR.

Reply:
I have added a benchmark test, using Rust `Bencher`. This is currently not being run by the runner. I can fix that in a separate PR. 

Here is the output: 
```
test tests::bench_wasm_handler ... bench:      25,608 ns/iter (+/- 976)
```
Is this the type of test you were thinking of?

Reply:
Nice! That time seems low compared to what I expected, so I guess that's good news. (I expected ~100ms roughly). Adding it to runner later on sounds good, although we should think about how to make sure we don't regress (just running the bench does not do much, it can easily climb up to 10s per request, and nothing will fail, unless we add additional logic to check that the time stays below a given threshold, e.g. 50ms for this function. Maybe Bencher already does this?

Reply:
BTW this crate seems to do something similar to what I'm thinking: https://bheisler.github.io/criterion.rs/book/user_guide/command_line_options.html#baselines

Reply:
That crate looks interesting. But I think for that to work on the CI builds we have to tweak some GitHub action scripts. For now I am calling `bench` directly on the `bencher` to be able to get the results and have assertions. I have also created #1933 to track benchmarking.